plugins/engines/optix/cuda/renderer/Phong.h

```diff
@@ -64,14 +71,75 @@ rtDeclareVariable(uint,              soft_shadows_enabled, , );
 rtDeclareVariable(float,             ambient_occlusion_strength, , );
 rtDeclareVariable(float4,            jitter4, , );
 
-rtBuffer<uchar4, 2>              output_buffer;
+// Volume
+rtBuffer<unsigned char> volumeData;
+rtDeclareVariable(uint3, volumeDimensions, , );
+rtDeclareVariable(float3, volumePosition, , );
+rtDeclareVariable(float3, volumeScale, , );
+rtDeclareVariable(float, volumeEpsilon, , );
+rtDeclareVariable(float, volumeDiag, , );
 
-rtDeclareVariable(optix::Ray, ray, rtCurrentRay, );
-rtDeclareVariable(float, t_hit, rtIntersectionDistance, );
-rtDeclareVariable(PerRayData_radiance, prd, rtPayload, );
-rtDeclareVariable(PerRayData_shadow,   prd_shadow, rtPayload, );
-rtDeclareVariable(unsigned int,  frame, , );
-rtDeclareVariable(uint2,         launch_index, rtLaunchIndex, );
+// Color map
+rtBuffer<float4> colorMap;
+rtDeclareVariable(uint, colorMapSize, , );
+rtDeclareVariable(float, colorMapMinValue, , );
+rtDeclareVariable(float, colorMapRange, , );
+
+rtBuffer<uchar4, 2> output_buffer;
+
+static __device__ float4 getVolumeValue()
+{
+    float3 pathColor = { 0.f, 0.f, 0.f };
+
+    float t = volumeEpsilon;
+    const float tMax = t_hit - volumeEpsilon + volumeDiag;
+
+    float pathAlpha = 0.f;
+    while( t < tMax && pathAlpha < 1.f )
+    {
+        float3 point = ( ray.origin + ray.direction * t ) / volumeScale - volumePosition;
+
+        if( point.x > 0.f && point.x < volumeDimensions.x &&
+            point.y > 0.f && point.y < volumeDimensions.y &&
+            point.z > 0.f && point.z < volumeDimensions.z )
+        {
+            ulong index = (ulong)(
+                (ulong)floor( point.x ) +
+                (ulong)floor( point.y ) * volumeDimensions.x +
+                (ulong)floor( point.z ) * volumeDimensions.x * volumeDimensions.y );
+
+            const unsigned char voxelValue = volumeData[ index ];
+
+            if( colorMap.size() != 0 )
+            {
+                const float normalizedValue =
+                    colorMapSize * ( voxelValue - colorMapMinValue ) /
```

Comment:
is colorMapSize not redundant with colorMap.size(), or do they just have confusingly similar names?